torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,746 @@
+                )
+                if i < len(self.grus) - 1:
+                    features = torch.cat([features, self._upsample2x(hidden_states[i + 1])], dim=1)
+
+                hidden_states[i] = self.grus[i](hidden_states[i], contexts[i], features)
+
+                # NOTE: For slow-fast gru, we dont always want to calculate delta depth for every call on UpdateBlock
+                # Hence we move the delta depth calculation to the RAFT-Stereo main forward
+
+        return hidden_states
+
+
+class MaskPredictor(raft.MaskPredictor):
+    """Mask predictor to be used when upsampling the predicted depth."""
+
+    # We add out_channels compared to raft.MaskPredictor
+    def __init__(self, *, in_channels: int, hidden_size: int, out_channels: int, multiplier: float = 0.25):
+        super(raft.MaskPredictor, self).__init__()
+        self.convrelu = Conv2dNormActivation(in_channels, hidden_size, norm_layer=None, kernel_size=3)
+        self.conv = nn.Conv2d(hidden_size, out_channels, kernel_size=1, padding=0)
+        self.multiplier = multiplier
+
+
+class CorrBlock1d(nn.Module):
+    """The row-wise correlation block.
+
+    Creates a row-wise correlation pyramid with ``num_levels`` levels from the outputs of the feature encoder,
+    and then indexes from this pyramid to create correlation features.
+    The "indexing" of a given centroid pixel x' is done by concatenating its surrounding row neighbours
+    within radius
+    """
+
+    def __init__(self, *, num_levels: int = 4, radius: int = 4):
+        super().__init__()
+        self.num_levels = num_levels
+        self.radius = radius
+
+        self.corr_pyramid: List[Tensor] = [torch.tensor(0)]  # useless, but torchscript is otherwise confused :')
+
+        self.out_channels = num_levels * (2 * radius + 1)
+
+    def build_pyramid(self, fmap1: Tensor, fmap2: Tensor):
+        """Build the correlation pyramid from two feature maps.
+
+        The correlation volume is first computed as the dot product of each pair (pixel_in_fmap1, pixel_in_fmap2) on the same row.
+        The last 2 dimensions of the correlation volume are then pooled num_levels times at different resolutions
+        to build the correlation pyramid.
+        """
+
+        if fmap1.shape != fmap2.shape:
+            raise ValueError(
+                f"Input feature maps should have the same shape, instead got {fmap1.shape} (fmap1.shape) != {fmap2.shape} (fmap2.shape)"
+            )
+        corr_volume = self._compute_corr_volume(fmap1, fmap2)
+
+        batch_size, h, w, num_channels, _ = corr_volume.shape
+        corr_volume = corr_volume.reshape(batch_size * h * w, num_channels, 1, w)
+        self.corr_pyramid = [corr_volume]
+        for _ in range(self.num_levels - 1):
+            corr_volume = F.avg_pool2d(corr_volume, kernel_size=(1, 2), stride=(1, 2))
+            self.corr_pyramid.append(corr_volume)
```

Comment:
From checking how this is used, it seems you are modifying the nn.Module on each batch forward and leave side-effects on the method. Can you try refactoring this (perhaps splitting in multiple modules?) to avoid keeping this information within the class? I haven't checked everything in details to tell you how to do it but I can further have a more detailed view if it helps.

Reply:
Yes, this is called on each batch forward once to update `self.corr_pyramid`. I think we can put this `self.corr_pyramid` on the RaftStereo class instead, do you think this is preferable?

Reply:
Ok, I have separate the module into two separate modules (one for corr_pyramid, another one for building corr_features)